New Project Flow using AbstractNewProjectStep

sonakshisaxena1 opened the pull request:
Enables the user to create a new project on clicking the "New Project" button through the Welcome Screen. 
![Screenshot from 2024-12-09 17-07-52](https://github.com/user-attachments/assets/82b656de-0917-4ec3-979e-bab9712f1e05)

emrekultursay commented on sherlock-platform/customization/src/com/google/sherlock/onboarding/EmptyProjectGenerator.java:
```diff
@@ -0,0 +1,121 @@
+  @Override
+  public @NotNull ValidationResult validate(@NotNull String baseDirPath) {
+    Path baseDir = Paths.get(baseDirPath);
+    if (!Files.exists(baseDir)) {
```
Should we invert this logic? If there is an existing directory, we probably want to reject it. We want Sherlock to be in charge of creating the directory (and all the files inside that directory).

Then we can remove the "expects an empty directory" section below.

emrekultursay commented on sherlock-platform/customization/src/com/google/sherlock/onboarding/EmptyProjectGenerator.java:
```diff
@@ -0,0 +1,121 @@
+/***
```
Can we (should we?) use Kotlin when we add new classes into the platform?

sonakshisaxena1 replied:
I believe if we are maintaining/implementing/extending an existing Java class, we should stick with Java for consistency and otherwise use Kotlin. Here, we implement `DirectoryProjectGenerator` hence I chose to continue with Java.

emrekultursay commented on sherlock-platform/customization/src/com/google/sherlock/onboarding/NewProjectAction.java:
```diff
@@ -0,0 +1,57 @@
+
+public class NewProjectAction extends AbstractNewProjectStep<NewProjectAction.MyProjectSettings> {
+
+  private static final Logger LOG = Logger.getInstance(NewProjectAction.class);
```
seems unused. Remove?

emrekultursay commented on sherlock-platform/customization/src/com/google/sherlock/onboarding/NewProjectAction.java:
```diff
@@ -0,0 +1,57 @@
+    // TODO: override getActions() for custom behavior
+  }
+
+  public static class MyProjectSettings {
```
Instead of `My`, can we call `SherlockProjectSettings`? 
// Or, can this be a private class called `ProjectSettings`?

emrekultursay commented on sherlock-platform/customization/src/com/google/sherlock/onboarding/ProjectSettingsStep.java:
```diff
@@ -0,0 +1,71 @@
+import com.intellij.platform.DirectoryProjectGenerator;
+import org.jetbrains.annotations.NotNull;
+
+import javax.swing.*;
```
No wildcard imports. I suggest changing your IntelliJ IDEA Java/Kotlin language settings to disable this.

emrekultursay commented on sherlock-platform/customization/src/com/google/sherlock/onboarding/EmptyProjectGenerator.java:
```diff
@@ -0,0 +1,121 @@
+    return ValidationResult.OK;
+  }
+
+  private void createProjectDirectory(@NotNull VirtualFile baseDir, @NotNull NewProjectAction.MyProjectSettings settings) {
```
Consider returning `projectDir` as `@NotNull VirtualFile` so that it can be passed to the next method addSampleFiles.

emrekultursay commented on sherlock-platform/customization/src/com/google/sherlock/onboarding/EmptyProjectGenerator.java:
```diff
@@ -0,0 +1,121 @@
+      LOG.error("Error creating project directory", e);
+      ApplicationManager.getApplication().invokeLater(() ->
+                                                        Messages.showErrorDialog("Error creating project directory: " + e.getMessage(), "Error")
+      );
```
I think we should rethrow the exception to prevent continuing into addSampleFiles.

sonakshisaxena1 commented:
This new commit updates most of the files that were pushed earlier, following the flow similar to that of Pycharm. This allows us to have more control over the New Project Wizard's UI as well as it's internal settings. The earlier flow restricted us to add "Project Name" in the creation window as well as assumed "empty settings" when it was tested using Breakpoints. This commit fixes all those. 

I have tried to use Kotlin when creating new classes, but also used Java at some places (like Pycharm) to ensure code consistency especially where the base classes are in Java. 

Some #TODOs have been included  in the  areas where I anticipate we'll update the flow in the future.

This is how the new UI looks - 


@emrekultursay apologies for such a large PR. I tried to reduce it as much as possible, but LMK if it becomes difficult to understand. PTAL at your convenience.

emrekultursay reviewed (commented):
Thanks for the progress! This PR is quite large, with a lot of new classes, so I can't claim that I understand which class does what. Maybe an explanation of which class serves what purpose (e.g., class docs) can help me understand better.  Here's a first round of code review.

sonakshisaxena1 commented on sherlock-platform/customization/src/com/google/sherlock/newProject/steps/SherlockGenerateProjectCallback.java:
```diff
@@ -0,0 +1,83 @@
+
+  @Nullable
+  public static Object computeProjectSettings(DirectoryProjectGenerator<?> generator) {
+    Object projectSettings = null;
```
Simplication would work but `getProjectSettings()` returns an `Object`, so I'll have to cast it to `SherlockNewProjectSettings` while returning. I'm fine with either so have updated according to your suggestion.

sonakshisaxena1 commented on sherlock-platform/customization/src/com/google/sherlock/newProject/steps/SherlockNewProjectStep.java:
```diff
@@ -0,0 +1,84 @@
+    protected @NotNull ProjectSettingsStepBase<SherlockNewProjectSettings> createProjectSpecificSettingsStep(
+      @NotNull DirectoryProjectGenerator<SherlockNewProjectSettings> projectGenerator,
+      @NotNull AbstractCallback<SherlockNewProjectSettings> callback) {
+      var npwGenerator = ObjectUtils.tryCast(projectGenerator, NewProjectWizardDirectoryGeneratorAdapter.class);
```
Yep, good point! Updated.

sonakshisaxena1 commented on sherlock-platform/customization/src/com/google/sherlock/newProject/steps/SherlockGenerateProjectCallback.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+
+    final Object settings = computeProjectSettings(generator);
```
Similar to the comment in Line 71- Provided return type by `getProjectSettings` is an `Object` hence casting is required.

sonakshisaxena1 commented on sherlock-platform/customization/src/com/google/sherlock/newProject/NewProjectWizardDirectoryGeneratorAdapter.kt:
```diff
@@ -0,0 +1,83 @@
+/***
```
Noted and updated.

sonakshisaxena1 commented on sherlock-platform/customization/src/com/google/sherlock/newProject/SherlockEmptyProjectGenerator.java:
```diff
@@ -0,0 +1,128 @@
+                                  final @NotNull Module module) {
+  }
+
+  public Object getProjectSettings() {
```
Sounds good!

emrekultursay replied:
This is still Object. Can it be concrete type?

sonakshisaxena1 commented on sherlock-platform/customization/src/com/google/sherlock/newProject/welcome/SherlockWelcome.kt:
```diff
@@ -0,0 +1,174 @@
+  }
+}
+
+internal object SherlockWelcomeGenerator {
```
Sounds good! I have modified this accordingly.

sonakshisaxena1 commented on sherlock-platform/customization/src/com/google/sherlock/newProject/steps/SherlockGenerateProjectCallback.java:
```diff
@@ -0,0 +1,83 @@
+    if (generationSettings == null) return null;
+    final DirectoryProjectGenerator generator = settings.getProjectGenerator();
+    final String location = FileUtil.expandUserHome(settings.getProjectLocation());
+    return AbstractNewProjectStep.doGenerateProject(null, location, generator, generationSettings);
```
You are right! It should be appended. I have fixed this now in `SherlockProjectSpecificSettingsStep.kt`

sonakshisaxena1 commented on sherlock-platform/customization/src/com/google/sherlock/newProject/steps/SherlockGenerateProjectCallback.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+
+    final Object settings = computeProjectSettings(generator);
```
Not yet. It's just a placeholder for dealing with settings in the future. I have added a `TODO` to clarify that it'll be used/updated eventually.

sonakshisaxena1 commented on sherlock-platform/customization/src/com/google/sherlock/newProject/steps/SherlockGenerateProjectCallback.java:
```diff
@@ -0,0 +1,83 @@
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class SherlockGenerateProjectCallback<T> extends AbstractNewProjectStep.AbstractCallback<T> {
```
I did try to change it but wasn't trivial as I ended up into a rabbit hole of errors and fixes. I would prefer to create a new issue catering this with a follow-up PR. I'm resolving this for now but LMK if you have a stronger opinion on getting this done right now. For what it's worth, this is exactly how PyCharm does it as well.

emrekultursay replied:
ACK in the context of  #32.

sonakshisaxena1 commented on sherlock-platform/customization/src/com/google/sherlock/newProject/steps/SherlockNewProjectStep.java:
```diff
@@ -0,0 +1,84 @@
+    @Override
+    public AnAction[] getActions(@NotNull List<? extends DirectoryProjectGenerator<?>> generators,
+                                 @NotNull AbstractCallback<SherlockNewProjectSettings> callback) {
+      generators = new ArrayList<>(generators);
```
Actually, this isn't even necessary until we decide to customise the actions. I added it to play around with the possibility of different generators and corresponding actions, and forgot to remove.

sonakshisaxena1 commented on sherlock-platform/customization/src/com/google/sherlock/newProject/SherlockEmptyProjectGenerator.java:
```diff
@@ -0,0 +1,128 @@
+import java.util.List;
+import java.util.function.Consumer;
+
+public abstract class SherlockEmptyProjectGenerator<T extends SherlockNewProjectSettings> extends DirectoryProjectGeneratorBase<T> {
```
Will be explored in a follow-up PR. I have created a new issue here - #32

sonakshisaxena1 commented on sherlock-platform/customization/src/com/google/sherlock/newProject/welcome/SherlockWelcome.kt:
```diff
@@ -0,0 +1,174 @@
+import org.jetbrains.annotations.CalledInAny
+import javax.swing.JPanel
+
+internal class SherlockWelcomeConfigurator : DirectoryProjectConfigurator {
```
Done.

sonakshisaxena1 commented:
I have now addressed the comments, removed some of the unnecessary code/files and included class documents to the best of my ability. Apologies again for such a large PR (I can split it into various PRs but it would be difficult for you to understand/see the cration flow until everything got merged due to the intertwined nature of the platform code! But happy to do so if docs don't help).
PTAL, thankyou! :)

emrekultursay commented on sherlock-platform/customization/src/com/google/sherlock/newProject/steps/SherlockProjectSpecificSettingsStep.kt:
```diff
@@ -0,0 +1,157 @@
+    val projectGenerator = myProjectGenerator
+    if (projectGenerator !is SherlockEmptyProjectGenerator<*>) return super.createBasePanel()
+
+    val nextProjectName = myProjectDirectory.get()
```
s/nextProjectName/nextProjectDir/

emrekultursay commented on sherlock-platform/customization/src/com/google/sherlock/newProject/steps/SherlockProjectSpecificSettingsStep.kt:
```diff
@@ -0,0 +1,157 @@
+  : ProjectSpecificSettingsStep<T>(projectGenerator, callback), DumbAware {
+  private val propertyGraph = PropertyGraph()
+  private val projectName = propertyGraph.property("")
+  private val projectLocation = propertyGraph.property("")
```
How exactly does your UI re-populate the same directory every time I restart IntelliJ? Is it a side-effect of "default project dir" being overwritten when we create a new project?

emrekultursay commented on sherlock-platform/customization/src/com/google/sherlock/newProject/steps/SherlockGenerateProjectCallback.java:
```diff
@@ -0,0 +1,110 @@
+   */
+  @Override
+  public void consume(@Nullable ProjectSettingsStepBase<T> step, @NotNull ProjectGeneratorPeer<T> projectGeneratorPeer) {
+    if (!(step instanceof ProjectSpecificSettingsStep settingsStep)) return;
```
Why not be more Sherlock-specific in this check?

```
if (!(step instanceof SherlockProjectSpecificSettingsStep settingsStep)) return;
```

emrekultursay commented on sherlock-platform/customization/src/com/google/sherlock/newProject/steps/SherlockGenerateProjectCallback.java:
```diff
@@ -0,0 +1,110 @@
+    //TODO: Update if we need to set WelcomeSettings before Project Generation.
+
+    final DirectoryProjectGenerator generator = settingsStep.getProjectGenerator();
+    if (generator instanceof SherlockEmptyProjectGenerator) {
```
Rename variable while checking for type?
```
if (generator instanceof SherlockEmptyProjectGenerator sherlockGenerator) { ...
}
```

Also, why not early return if it's not of type `Sherlock`? (similar to line 46 above)

emrekultursay commented on sherlock-platform/customization/src/com/google/sherlock/newProject/steps/SherlockGenerateProjectCallback.java:
```diff
@@ -0,0 +1,110 @@
+    }
+
+
+    final SherlockNewProjectSettings settings = computeProjectSettings(generator, settingsStep);
```
Looking at lines 65-68, I think Kotlin's explicit notnull/nullable representation would make it more readable and less error prone.

emrekultursay commented on sherlock-platform/customization/src/com/google/sherlock/newProject/steps/SherlockGenerateProjectCallback.java:
```diff
@@ -0,0 +1,110 @@
+  private static Project generateProject(@NotNull final ProjectSettingsStepBase settings, @Nullable Object generationSettings) {
+    if (generationSettings == null) return null;
+    final DirectoryProjectGenerator generator = settings.getProjectGenerator();
+    final String location = FileUtil.expandUserHome(settings.getProjectLocation());
```
`expandUserHome` was already called inside settingsStep. It's not needed here.

emrekultursay commented on sherlock-platform/customization/src/com/google/sherlock/newProject/SherlockEmptyProjectGenerator.java:
```diff
@@ -0,0 +1,141 @@
+   * @param settings The project settings.
+   * @param module   The main module of the project.
+   */
+  protected void configureProject(final @NotNull Project project,
```
Why not leave it abstract with no implementation (instead of concrete with empty body)? Then subclasses won't have to call super.configureProject().

emrekultursay commented on sherlock-platform/customization/src/com/google/sherlock/newProject/welcome/SherlockWelcome.kt:
```diff
@@ -0,0 +1,197 @@
+            .onUiThread(ModalityState.nonModal())
+            .expireWith(SherlockPluginDisposable.getInstance(project))
+            .submit {
+              val fileToChoose = (file ?: firstUserFile(project, baseDir, module)) ?: return@submit
```
When I try the PR, I always get `null` here. I think that's because the project is currently in-memory, but not flushed into disk, so the file system doesn't find any files. Maybe we need some `project.save()` before we do this? 

Also, for the same reason, the Project View shows empty when I create a project, but when I hit `Cmd+S`, it  refreshes the tree and shows the files.

emrekultursay reviewed (commented):
Thanks for adding the comments, and the doc was also very useful.

sonakshisaxena1 commented on sherlock-platform/customization/src/com/google/sherlock/newProject/SherlockEmptyProjectGenerator.java:
```diff
@@ -0,0 +1,141 @@
+    return null;
+  }
+
+  public @Nullable JPanel extendBasePanel() throws ProcessCanceledException {
```
IntelliJ is not able to detect some methods. This gets overidden in `SherlockBaseProjectGenerator`. Also, I want to keep this as it can be used to extend the base panel to include further settings for Sherlock Empty Project like including samples etc.

emrekultursay replied:
> IntelliJ is not able to detect some methods' usages but this gets overidden in SherlockBaseProjectGenerator.

Yes, it's overridden, but neither this base method, nor the override in the subclass seems to be *called* anywhere. Are they called?

> Also, I want to keep this as I see it being used to extend the base panel (eventually)...

In general, I discourage adding unused functionality just because they may (or will) be used in the future. I'm OK with adding a call to it, but leaving the body as `// TODO: Add customizations to the panel here.`.  However, if it's not used in the current version/commit, then it's proven dead code, and shall be deleted.

sonakshisaxena1 commented on sherlock-platform/customization/src/com/google/sherlock/newProject/steps/SherlockProjectSpecificSettingsStep.kt:
```diff
@@ -0,0 +1,157 @@
+        projectNameFiled = textField()
+          .bindText(projectName)
+          .validationOnInput {
+            val validationResult = projectGenerator.validate(getProjectLocation())
```
Yes, I don't think we should have both. Honestly, I don't really know how "Create from existing sources" is different than "Open" as it ultimately opens the existing directory. My follow-up on this PR is to figure out and fix the opening of the projects. 

ACK for validation.

sonakshisaxena1 commented on sherlock-platform/customization/resources/META-INF/SherlockPlatformCustomization.xml:
```diff
@@ -1,14 +1,35 @@
+    <extensions defaultExtensionNs="com.intellij">
+      <directoryProjectConfigurator implementation="com.google.sherlock.newProject.welcome.SherlockWelcomeConfigurator"
+                                    id="welcome" order="last"/>
+      <statistics.counterUsagesCollector implementationClass="com.google.sherlock.newProject.welcome.SherlockWelcomeCollector"/>
```
This basically records user or internal IDE actions. Can be useful for us to understand user onboarding behaviour and for testing. We don't need this right now, but I thought it's nice to have! I have no strong opinion on keeping/removing it. What do you think?

emrekultursay replied:
I'd suggest deleting the class and commenting out this line 19 with a note. We can add it when we need it.

emrekultursay commented on sherlock-platform/customization/src/com/google/sherlock/newProject/steps/SherlockGenerateProjectCallback.java:
```diff
@@ -0,0 +1,101 @@
+   * @return The newly created project.
+   */
+  @NotNull
+  private static Project generateProject(@NotNull final ProjectSettingsStepBase settings, @Nullable Object generationSettings) {
```
s/Object/SherlockNewProjectSettings/

sonakshisaxena1 replied:
ACK.

emrekultursay commented on sherlock-platform/customization/resources/messages/SherlockWelcomeBundle.properties:
```diff
@@ -0,0 +1,16 @@
+#Copyright 2024 Google LLC
```
Unused file. Can be added later. Same for `SherlockBundle.kt`

sonakshisaxena1 commented on sherlock-platform/customization/src/com/google/sherlock/newProject/steps/ProjectSpecificSettingsStep.java:
```diff
@@ -0,0 +1,41 @@
+
+  @Override
+  protected JPanel createAndFillContentPanel() {
+    if (myProjectGenerator instanceof SherlockEmptyProjectGenerator) {
```
Agree, updated.

sonakshisaxena1 commented on sherlock-platform/customization/src/com/google/sherlock/newProject/steps/SherlockProjectSpecificSettingsStep.kt:
```diff
@@ -0,0 +1,160 @@
+  projectGenerator: DirectoryProjectGenerator<T>,
+  callback: AbstractNewProjectStep.AbstractCallback<T>,
+)
+  : ProjectSpecificSettingsStep<T>(projectGenerator, callback), DumbAware {
```
I have now removed `ProjectSpecificSettingStep` so now this class directly inherits from `ProjectSettingsStepBase` instead.

sonakshisaxena1 commented on sherlock-platform/customization/src/com/google/sherlock/newProject/steps/SherlockNewProjectStep.java:
```diff
@@ -0,0 +1,55 @@
+    protected @NotNull ProjectSettingsStepBase<SherlockNewProjectSettings> createProjectSpecificSettingsStep(
+      @NotNull DirectoryProjectGenerator<SherlockNewProjectSettings> projectGenerator,
+      @NotNull AbstractCallback<SherlockNewProjectSettings> callback) {
+      return new SherlockProjectSpecificSettingsStep<>(projectGenerator, callback);
```
ACK.

emrekultursay commented on sherlock-platform/customization/resources/META-INF/SherlockPlatformCustomization.xml:
```diff
@@ -1,14 +1,35 @@
+    <extensions defaultExtensionNs="com.intellij">
+      <directoryProjectConfigurator implementation="com.google.sherlock.newProject.welcome.SherlockWelcomeConfigurator"
+                                    id="welcome" order="last"/>
+      // Include statistics.counterUsagesCollector for recording user or IDE actions.
```
In xml, use: `<!-- comment -->`

emrekultursay commented on sherlock-platform/customization/src/com/google/sherlock/newProject/steps/SherlockBaseProjectGenerator.kt:
```diff
@@ -0,0 +1,59 @@
+   *
+   * @return The logo for the project generator.
+   */
+  //TODO: Update when we have a Logo
```
nit: this TODO should be inside the body of the function. (also applies to a comment in the next file)

emrekultursay commented on sherlock-platform/customization/src/com/google/sherlock/newProject/ui/MainPartUiCustomizer.kt:
```diff
@@ -11,18 +11,19 @@
+ */
+interface MainPartUiCustomizer {
+  /**
+   * Adds a panel under the checkbox section.
```
What's a "checkbox section"?

emrekultursay commented on sherlock-platform/customization/src/com/google/sherlock/newProject/steps/SherlockProjectSpecificSettingsStep.kt:
```diff
@@ -0,0 +1,161 @@
+      row("") {
+        comment("", maxLineLength = 60).bindText(locationHint)
+      }
+      val uiCustomizer = projectGenerator.mainPartUiCustomizer
```
AFAIU, this customizer is not doing anything yet.  Maybe leave MainPartUiCustomizer out of this change, and leave a `// TODO: add UI customizer` here? (I'll leave the decision to you).
